web/explorer/src/utils/rdocParser.js

```diff
@@ -3,69 +3,219 @@
+        treeData.push(ruleNode);
+        index++;
+    }
+    return treeData;
+}
 
-        if (isFileScope) {
-            // The scope for the rule is a file, so we don't need to show the match location address
-            ruleNode.children = limitedMatches.map((match, matchIndex) => {
-                return parseNode(match[1], `${index}-${matchIndex}`, rules, rule.meta.lib, layout);
-            });
-        } else {
-            // This is not a file-level match scope, we need to create intermediate nodes for each match
-            ruleNode.children = limitedMatches.map((match, matchIndex) => {
-                const matchKey = `${index}-${matchIndex}`;
-                const matchNode = {
-                    key: matchKey,
-                    data: {
-                        type: "match location",
-                        name:
-                            flavor === "static"
-                                ? `${rule.meta.scopes.static} @ ` + formatAddress(match[0])
-                                : getProcessName(layout, match[0])
-                    },
-                    children: [parseNode(match[1], `${matchKey}`, rules, rule.meta.lib, layout)]
-                };
-                return matchNode;
-            });
+/**
+ * Preprocesses the layout to create efficient lookup maps
+ * @param {Object} layout - The layout object from rdoc JSON data
+ * @returns {Object} An object containing lookup maps for calls, threads, and processes
+ */
+function preprocessLayout(layout) {
+    const processMap = new Map();
+    const threadMap = new Map();
+    const callMap = new Map();
+
+    if (layout && layout.processes) {
+        for (const process of layout.processes) {
+            if (process.address && process.address.type === "process" && process.address.value) {
+                const [ppid, pid] = process.address.value;
+                processMap.set(`${ppid}-${pid}`, process);
+
+                if (process.matched_threads) {
+                    for (const thread of process.matched_threads) {
+                        if (thread.address && thread.address.type === "thread" && thread.address.value) {
+                            const [, , tid] = thread.address.value;
+                            threadMap.set(`${ppid}-${pid}-${tid}`, thread);
+
+                            if (thread.matched_calls) {
+                                for (const call of thread.matched_calls) {
+                                    if (call.address && call.address.type === "call" && call.address.value) {
+                                        const [, , , callId] = call.address.value;
+                                        callMap.set(`${ppid}-${pid}-${tid}-${callId}`, call);
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    return { processMap, threadMap, callMap };
+}
```

Comment:
@williballenthin @mr-tz I did some profiling yesterday on my code to identify where the `parseRules` spends most of the time and to my surprise parsing the 80MB rdoc is fairly quick. The issue lays in retrieving the call name (e.g. `explorer.exe{pid:1000,...}`)  [which has O(n**3)](https://github.com/mandiant/capa/blob/22d20ed2b8b830fbe659ccec2984cf35c2fd1896/capa/render/verbose.py#L89C1-L105C1). In my case, I refactored `parseRules` function to pre-compute static maps and use those for lookup which proved to be faster in my case. As a result, I am able to render 300 matches (instead of 25) per rule in a relatively short time.

Reply:
You can preview loading 300 matches using: https://s-ff.github.io/capa/explorer/#/?rdoc=https://raw.githubusercontent.com/mandiant/capa-testfiles/master/rd/0000a65749f5902c4d82ffa701198038f0b4870b00a27cfca109f8f933476d82.json.gz

Reply:
awesome!